File: Array Sorting Algorithms/Radix Sort/src/src-java/radix-sort.java
import java.util.Arrays;

public class RadixSort {
    public static void radixSort(int[] arr) {
        int max = Arrays.stream(arr).max().getAsInt();
        int exp = 1;

        while (max / exp > 0) {
            countingSortRadix(arr, exp);
            exp *= 10;
        }
    }

    public static void countingSortRadix(int[] arr, int exp) {
        int n = arr.length;
        int[] output = new int[n];
        int[] countingArr = new int[10];

        for (int num : arr) {
            countingArr[(num / exp) % 10]++;
        }

        for (int i = 1; i < 10; i++) {
            countingArr[i] += countingArr[i - 1];
        }

        for (int i = n - 1; i >= 0; i--) {
            output[countingArr[(arr[i] / exp) % 10] - 1] = arr[i];
            countingArr[(arr[i] / exp) % 10]--;
        }

        System.arraycopy(output, 0, arr, 0, n);
    }

    // Example usage:
    public static void main(String[] args) {
        int[] arr = {170, 45, 75, 90, 802, 24, 2, 66};
        radixSort(arr);
        System.out.println("Sorted array: " + Arrays.toString(arr));
    }
}

File: array-sorting-algorithms/radix-sort/src/src-python/radix-sort.py
def radix_sort(arr):
    max_value = max(arr)
    exp = 1

    while max_value // exp > 0:
        counting_sort_radix(arr, exp)
        exp *= 10

def counting_sort_radix(arr, exp):
    n = len(arr)
    output = [0] * n
    counting_arr = [0] * 10

    for num in arr:
        index = num // exp
        counting_arr[index % 10] += 1

    for i in range(1, 10):
        counting_arr[i] += counting_arr[i - 1]

    i = n - 1
    while i >= 0:
        index = arr[i] // exp
        output[counting_arr[index % 10] - 1] = arr[i]
        counting_arr[index % 10] -= 1
        i -= 1

    for i in range(n):
        arr[i] = output[i]

# Example usage:
arr = [170, 45, 75, 90, 802, 24, 2, 66]
radix_sort(arr)
print("Sorted array:", arr)
